### PR TITLE
feat: Implement and install new ABCI Query for transaction inclusion proofs 

### LINKED ISSUE #698: Create a custom query route for tendermint to use to create transaction inclusion proofs
Currently, when creating a tx inclusion proof in celestia-core, we have to encode a portion of the block data into rows. This is necessary to create the nmt tree(s) used to create the proof(s). It's problematic not only because we don't want to have any application logic in core, but in order to fully test that code we have to have a lot of other application specific logic (malleated txs, non-interactive defaults/other square layout rules).

To fix this, we'd like to move all share encoding logic to the app, but this means we can't do tx inclusion proofs in core like normal tendermint. We can get around this by creating a [custom query route](https://github.com/celestiaorg/cosmos-sdk/blob/368049e57edb435ffc58a7e7afedab7e54d83776/baseapp/abci.go#L443-L445), and calling that route each time we want to create a tx inclusion proof in tendermint.

This will involve:
- [x] #699
- [ ] #708
- [x] https://github.com/celestiaorg/celestia-core/issues/851

### PR DESCRIPTION
**Pull request by evan-forbes**
This PR introduces a new custom ABCI query that generates the ABCI inclusion proof for a transaction of a specific index. 

Note that this PR does not fix the tx inclusion proof logic to follow the non-interactive defaults, as we haven't merged those yet. It does however alter the integration test to require that transactions include proofs, so we are testing the new ABCI query method.

blocked by https://github.com/celestiaorg/celestia-core/pull/859
closes #698

### CODE REVIEW

**Comment by evan-forbes** [comment]
NOTE: we probably should not merge this until we cut a new release of tendermint with the new query usuage and replace that tag in the go.mod

File: app/test/block_size_test.go
Comment (rootulp) [inline app/test/block_size_test.go:131]: [question] is this comment still necessary after we pass `prove=true` below?

```suggestion
```

File: app/split_shares.go
Comment (evan-forbes) [inline app/split_shares.go:117]: quite a bit of code had to be changed in the shareSplitter in order to update it to work with the local copies of the share encoding code, as that was never done before as I assumed we'd merge non-interactive defaults code first :sweat_smile: 

would prefer not create a separate PR for this, but I understand if its something we have to do

File: app/test/block_size_test.go
Comment (evan-forbes) [inline app/test/block_size_test.go:312]: this change requires that we include a tx inclusion proof when querying for a tx

File: pkg/prove/querier.go
Comment (rootulp) [inline pkg/prove/querier.go:22]: [question] which fragment of the path is correct `tx-inclusion-proof` or `txInclusionProof`?

File: app/test/block_size_test.go
Comment (evan-forbes) [inline app/test/block_size_test.go:138]: Now we're requiring that all transactions queried for during the integration test have valid inclusion proofs.

**Review by rootulp — APPROVED** [review]

File: pkg/prove/querier.go
Comment (evan-forbes) [reply pkg/prove/querier.go:22]: good catch! 28da36536c35c8e488f108ee77b72dcb9b5d1b1c

File: app/test/block_size_test.go
Comment (evan-forbes) [reply app/test/block_size_test.go]: good catch d1f615293bfad076cda6913880450b160842c383

**Review by rootulp — APPROVED** [review]
:shipit:

**Comment by evan-forbes** [comment]
still blocked by celestia-core, and need to update to the new release before we can merge, so converting to a draft until then

File: pkg/prove/querier.go
Comment (rootulp) [inline pkg/prove/querier.go:14]: [question] does this not need the `custom/` prefix that is used in https://github.com/celestiaorg/celestia-core/pull/859/files#diff-6572dddc535b4d2c98746d71dde97840af1328797e193db28376939a320be8f8R48 ?

**Review by rootulp — APPROVED** [review]

File: pkg/prove/querier.go
Comment (evan-forbes) [reply pkg/prove/querier.go:14]: per comment in core, this one needs to not include the path as it will be removed/added by the cosmos-sdk

**Review by rootulp — APPROVED** [review]

File: pkg/prove/querier.go
Comment (SweeXordious) [inline pkg/prove/querier.go:18]: [typo]
```suggestion
// proved must be appended to the path. The marshalled bytes of the transaction
```

File: pkg/prove/querier.go
Comment (SweeXordious) [inline pkg/prove/querier.go:26]: [Non Blocking][Cosmetics]
```suggestion
		return nil, fmt.Errorf("expected query path length: 1 actual: %d ", len(path))
```

File: pkg/prove/querier.go
Comment (SweeXordious) [inline pkg/prove/querier.go:45]: [Unrelated][Question]
In the docs of `TxInclusion(...)`:
```
// TxInclusion uses the provided block data to progressively generate rows
// of a data square, and then using those shares to creates nmt inclusion proofs
// It is possible that a transaction spans more than one row. In that case, we
// have to return two proofs.
func TxInclusion(codec rsmt2d.Codec, data types.Data, txIndex uint64) (types.TxProof, error) {
```
I checked the implementation, and if I understand right, it will generate a proof for each row a transaction spans on?
If so, why does it state in the docs:
```
In that case, we have to return two proofs.
```
and also:
```
var proofs []*tmproto.NMTProof  //nolint:prealloc // rarely will this contain more than a single proof
```

**Review by SweeXordious — APPROVED** [review]
Awesome stuff :rocket:
Only questions to understand more and cosmetics.

File: pkg/prove/querier.go
Comment (evan-forbes) [reply pkg/prove/querier.go:45]: They seem to match to me, but perhaps we can dig a bit deeper and clarify them. not sure I'm understanding the question properly, mind elaborating? which part doesn't match or make sense?

Comment (SweeXordious) [reply pkg/prove/querier.go:45]: The part I am not getting is the **two** proofs. Why not any number of proofs? Will the number of proofs be the same as the number of rows?

Comment (evan-forbes) [reply pkg/prove/querier.go:45]: true! yeah, we can change that, but all normal txs will only ever result in two proofs. You're correct tho that the implementation can handle more than two if that ever does happen (submitting a tx with a giiiiaaaannnnnntttttt memo maybe?)

Comment (evan-forbes) [reply pkg/prove/querier.go:45]: https://github.com/celestiaorg/celestia-app/pull/708/commits/26b8baa00ee372f609e8de7260b010f65c94b696

Comment (SweeXordious) [reply pkg/prove/querier.go:45]: Awesome,  thanks a lot

**Review by SweeXordious — APPROVED** [review]
:shipit:

File: pkg/prove/proof.go
Comment (rootulp) [inline pkg/prove/proof.go:19]: [grammar] honestly may not be worth accepting this suggestion b/c it will dismiss approvals

```suggestion
// have to return more than one proof.
```

**Review by rootulp — APPROVED** [review]

File: pkg/prove/proof.go
Comment (evan-forbes) [reply pkg/prove/proof.go:19]: :grimacing: degen mergingggggggg

Comment (evan-forbes) [reply pkg/prove/proof.go:19]: I will sneak this into a different PR tho lol

**Pull request merged**
